main.js

```diff
@@ -0,0 +1,616 @@
+  client = null;
+  if (txtNick.value.trim() != '' && txtPass.value.trim() != '') {
+    //console.log('hi');
+    client = new TwitchClient({
+      Nick: txtNick.value,
+      Pass: txtPass.value,
+      Channels: txtChannel.value,
+      Debug: debug,
+    });
+  } else {
+    client = new TwitchClient({
+      Channels: txtChannel.value,
+      Debug: debug,
+    });
+  }
+
+  client.onRoomstate = function(channel, settings) {
+    //console.log(channel, settings['room-id']);
+    GetChannelBadges(settings['room-id']);
+  };
+
+  client.onPrivmsg = function(user, channel, message, userData, rawMessage) {
+    if (message_history[message_history.length - 1] == rawMessage) return;
+
+    if (message_history.length == message_history_length)
+      message_history.shift();
+
+    message_history.push(rawMessage);
+
+    var p = ParseMessage(user, message, userData);
+
+    //go through each module and append the message
+    //TODO: message filtering
+    $('.module').each(function() {
+      var id = $(this).attr('id');
+      var $content = $(this).find('.content');
+
+      var el = $content[0],
+        scroll = false;
+
+      if (el.clientHeight + el.scrollTop >= el.scrollHeight - 100)
+        scroll = true;
+
+      var disp = false;
+
+      //plebs
+      if (config[id].Pleb && !userData['subscriber']) disp = true;
+
+      //subs
+      if (config[id].Sub && userData['subscriber']) disp = true;
+
+      //mods
+      if (config[id].Mod && userData['mod']) disp = true;
+
+      if (config[id].Bits && userData['bits']) disp = true;
+
+      for (s of config[id].IncludeUser) {
+        if (user.toLowerCase() == s.toLowerCase()) disp = true;
+      }
+
+      for (s of config[id].IncludeKeyword) {
+        if (message.toLowerCase().indexOf(s.toLowerCase()) > -1) disp = true;
+      }
+
+      for (s of config[id].ExcludeUser) {
+        if (user.toLowerCase() == s.toLowerCase()) disp = false;
+      }
+
+      for (s of config[id].ExcludeStartsWith) {
+        if (message.toLowerCase().indexOf(s.toLowerCase()) == 0) disp = false;
+      }
+
+      if (disp) $content.append(p);
+
+      if (scroll) el.scrollTop = el.scrollHeight;
+    });
+  };
+
+  client.onUsernotice = function(message) {
+    if (message_history.length == message_history_length)
+      message_history.shift();
+
+    message_history.push(message);
+
+    //console.log(message);
+  };
+}
+
+function ParseMessage(user, message, userData) {
+  //use specified color if there is one; if not, pick a random one from the defaults and store it
+  var user_col = userData['color'];
+  if (user_col == '') {
+    if (!user_undefined_colors[user]) {
+      user_undefined_colors[user] = GetRandomColor();
+    }
+
+    user_col = user_undefined_colors[user];
+  }
+
+  message = EscapeHtml(message);
+
+  var message_col = '';
+  if (message.indexOf('ACTION') == 0) {
+    message_col = `color: ${user_col}`;
+    message = message.substring(8, message.length - 1);
+  }
+
```

Comment:
Missing chat message filtering from 985d77439aa0f5c4e8861a05a963024b27ec20c8
```suggestion
  message = message.replace(/</g, '&lt;').replace(/>/g, '&gt');
```

Reply:
Does this change reside within the if statement on line 396 or outside of it? Unclear from suggested change overview.

Reply:
Just outside it. Now that you mention it, I need to fix the indentation on that line, so I'll do that now.